File: middlewares/thousandsMiddleware.js
const convertTriplet = require('../extensions/convertTriplet');
const { thousandWord } = require('../extensions/pluralize');

function thousandsMiddleware(req, res, next) {
  const thousands = Math.floor(req.num / 1000);

  if (thousands > 0) {
    const thousandWords = convertTriplet(thousands, true);
    req.words.push(...thousandWords, thousandWord(thousands));
  }

  req.num = req.num % 1000;
  next();
}

module.exports = thousandsMiddleware;

File: Middlewares/ThousandsMiddleware.cs
using Microsoft.AspNetCore.Http;
using NumberInterpreter.Extensions;

namespace NumberInterpreter.Middlewares;

public class ThousandsMiddleware
{
    private readonly RequestDelegate _next;

    public ThousandsMiddleware(RequestDelegate next)
    {
        _next = next;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        var num = (int)context.Items["Num"]!;
        var words = (List<string>)context.Items["Words"]!;

        var thousands = num / 1000;

        if (thousands > 0)
        {
            words.AddRange(thousands.ConvertTriplet(true));
            words.Add(thousands.ThousandWord());
        }

        context.Items["Num"] = num % 1000;

        await _next(context);
    }
}
